sdk/python/feast/pyspark/historical_feature_retrieval_job.py

```diff
@@ -1,19 +1,318 @@
+import argparse
 import json
-from datetime import timedelta
-from typing import Any, Dict, List
+from datetime import datetime, timedelta
+from typing import Dict, List, NamedTuple, Optional
 
-from pyspark import SparkFiles
 from pyspark.sql import DataFrame, SparkSession, Window
 from pyspark.sql.functions import col, expr, monotonically_increasing_id, row_number
 
 
+class EntitySource(NamedTuple):
+    """
+    Source for the entity dataframe. The values for the `format` and `path` should be
+    recognizable by the Spark cluster where the job is going to run on. For example, s3a
+    connector is required for Amazon S3 path. Spark avro package should be on the class
+    path of the spark cluster if the format is `avro`.
+
+    Attributes:
+        format (str): File format.
+        path (str): uri to the file source.
+        timestamp_column (str): Column representing the event timestamp.
+        mapping (Dict[str, str]): Optional. If present, the source column will be renamed
+            based on the mapping.
+        options (Dict[str, str]): Spark read options.
+    """
+
+    format: str
+    path: str
+    timestamp_column: str
+    mapping: Dict[str, str] = {}
+    options: Dict[str, str] = {}
+
+
+class FeatureTableSource(NamedTuple):
+    """
+    Source for the Feature Table dataframe. The values for the `format` and `path` should be
+    recognizable by the Spark cluster where the job is going to run on. For example, s3a
+    connector is required for Amazon S3 path. Spark avro package should be on the class
+    path of the spark cluster if the format is `avro`.
+
+    Attributes:
+        format (str): File format.
+        path (str): uri to the file source.
+        timestamp_column (str): Column representing the event timestamp.
+        created_timestamp_column (str): Column representing the creation timestamp. Not required for entity.
+        mapping (Dict[str, str]): Optional. If present, the source column will be renamed
+            based on the mapping.
+        options (Dict[str, str]): Spark read options.
+    """
+
+    format: str
+    path: str
+    timestamp_column: str
+    created_timestamp_column: str
+    mapping: Dict[str, str] = {}
+    options: Dict[str, str] = {}
+
+
+class Field(NamedTuple):
+    """
+    Defines name and type for Feast entities and features.
+
+    Attributes:
+        name (str): Field name.
+        type (str): Feast type name.
+    """
+
+    name: str
+    type: str
+
+    @property
+    def spark_type(self):
```

Comment:
should this be `to_spark_type`?

Reply:
`spark_type` would be more natural, since the intended way to call them would be:

`>>> entity = Field(name="driver", type="int32")`
`>>> entity.type`
"int32"
`>>>  entity.spark_type`
"int"

Reply:
Ok. Doesnt seem wrong per say, but is there a reason for this approach? Isn't it more pythonic to use getters and setters, like `entity.to_spark_type()`?